### PR TITLE
push utf8 comparison down into ValueReader

### PR DESCRIPTION
**Pull request by richardstartin**
This pushes the unsigned comparison operation down into the `ValueReader` so that a string doesn't need to be materialised to do the comparison. For `FixedByteValueReaderWriter`, which has showed up as a CPU bottleneck in pretty much every profile of Pinot query evaluation I have looked at, the work done to find where the string ends within the fixed width bucket before copying the string can be skipped.

### CODE REVIEW

**Comment by richardstartin** [comment]
Here are some preliminary numbers from my laptop (these need taking with a pinch of salt but look very promising)

master:

```
Benchmark                                                            Mode  Cnt     Score      Error  Units
BenchmarkStringVarLengthDictionary.fixedStringDictionaryGet          avgt    5   253.967 ±  593.088  ms/op
BenchmarkStringVarLengthDictionary.fixedStringDictionaryIndexOf      avgt    5  8123.103 ± 2528.026  ms/op
BenchmarkStringVarLengthDictionary.varLengthStringDictionaryGet      avgt    5   224.049 ±  312.375  ms/op
BenchmarkStringVarLengthDictionary.varLengthStringDictionaryIndexOf  avgt    5  8071.103 ± 2135.835  ms/op
```

branch
```
Benchmark                                                            Mode  Cnt     Score      Error  Units
BenchmarkStringVarLengthDictionary.fixedStringDictionaryGet          avgt    5   253.967 ±  593.088  ms/op
BenchmarkStringVarLengthDictionary.fixedStringDictionaryIndexOf      avgt    5  8123.103 ± 2528.026  ms/op
BenchmarkStringVarLengthDictionary.varLengthStringDictionaryGet      avgt    5   224.049 ±  312.375  ms/op
BenchmarkStringVarLengthDictionary.varLengthStringDictionaryIndexOf  avgt    5  8071.103 ± 2135.835  ms/op
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReader.java [hunk 1]
```diff
@@ -20,6 +20,9 @@
 
 import java.io.Closeable;
 import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
```
Comment (richardstartin) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReader.java:25]: there might be a better place for the shared logic than a static on this interface, which would keep `PinotDataBuffer` out of this interface. I'm open to suggestions.

**Comment by richardstartin** [comment]
It seems there are some corner cases to address so benchmarking is premature

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java [hunk 2]
```diff
@@ -259,8 +259,7 @@ protected int binarySearch(byte[] value) {
 
     while (low <= high) {
       int mid = (low + high) >>> 1;
-      byte[] midValue = _valueReader.getBytes(mid, _numBytesPerValue);
-      int compareResult = ByteArray.compare(midValue, value);
+      int compareResult = _valueReader.compareUnsignedBytes(mid, _numBytesPerValue, value);
```
Comment (richardstartin) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java:262]: this introduces a bug as things stand as it will compare to the prefix before the first zero in the stored data, I will fix this soon and add some testing around this scenario.

Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java]: The existing logic here appears to be broken for fixed length `FixedByteValueReaderWriter`: it compares `value` with the _padded_ stored value retrieved by `ValueReader.getBytes`. This comparison will break ties using the lengths when mismatches can't be found, meaning the comparison will never yield zero unless `value` is also padded. I can only find testing for this method with byte arrays with the same size, so this problem would not arise in tests, but I can't find where the padding required to make this work with arbitrary length byte arrays is taking place.

I decided to roll this change back as addressing this problem is bigger than the scope of the change I wanted to make, and `STRING` columns are much more common than `BYTES`. I refactored the code to make adding an unsigned comparison trivial once this has been resolved.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        }
+        byte ourByte = dataBuffer.getByte(startOffset + mismatchPosition);
+        if (ourByte == 0) {
+            // then we are storing the empty string
+            return -1;
+        }
+        byte theirByte = bytes[mismatchPosition];
+        if (theirByte == 0) {
+            return 1;
+        }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:89]: I don't think we need these special handling since they are not very common case. We may change `first > 0` to `first >= 0` to handle the 0 byte in `compareUtf8()`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    private static int compareUtf8(PinotDataBuffer ourBuffer, long ourStartOffset,
+                           ByteBuffer theirBuffer, int mismatchPosition) {
+        char ours1 = '?';
+        char ours2 = '?';
+        char theirs1 = '?';
+        char theirs2 = '?';
+
+        // 1. seek backwards from mismatch position in each buffer to find start of each utf8 sequence
+        long ourPosition = ourStartOffset + mismatchPosition;
+        while (ourPosition > ourStartOffset && isUtf8Continuation(ourBuffer.getByte(ourPosition))) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:104]: The boundary check is redundant. We can also move both our and their position together for better performance

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                ours1 = (char) (first & 0xFF);
+            } else if (control < 0xE0) {
+                ours1 = decode(first, ourBuffer.getByte(ourPosition + 1));
+            } else if (control == 0xE0) {
+                ours1 = decode(first, ourBuffer.getByte(ourPosition + 1),
+                        ourBuffer.getByte(ourPosition + 2));
+            } else {
+                int codepoint = decode(first, ourBuffer.getByte(ourPosition + 1),
+                        ourBuffer.getByte(ourPosition + 2), ourBuffer.getByte(ourPosition + 3));
+                if (Character.isValidCodePoint(codepoint)) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:125]: Does the logic handle the case when the code point is invalid for one value but valid for the other? I feel using `'?'` as the missing character can cause unexpected behavior in such case.

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/readerwriter/ValueReaderComparisonTest.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class ValueReaderComparisonTest {
+    @DataProvider
+    public static Object[] text() {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/readerwriter/ValueReaderComparisonTest.java:45]: (Code format) Please fix the indentation

**Review by Jackie-Jiang — COMMENTED** [review]
This is great! I tried the similar approach, but blocked on the utf-8 to char conversion. Great job! Mostly minor comments

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    private ValueReaderComparisons() {
+    }
+
+    private static int mismatch(PinotDataBuffer dataBuffer, long startOffset, int length, ByteBuffer buffer) {
+        boolean littleEndian = dataBuffer.order() == ByteOrder.LITTLE_ENDIAN;
+        if (littleEndian) {
+            buffer.order(ByteOrder.LITTLE_ENDIAN);
+        }
+        int limit = Math.min(length, buffer.limit());
+        int loopBound = limit & ~0x3F;
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:36]: Good catch, this is a debugging artifact.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+    static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded,
+                           byte[] bytes) {
+        // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+        ByteBuffer buffer = ByteBuffer.wrap(bytes);
+        int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+        if (mismatchPosition == -1) {
+            // need to figure out whether the unpadded string is longer than the parameter or not
+            if (padded && bytes.length < length) {
+                if (bytes.length == 0) {
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:66]: I thought so too, but there are pre-existing test cases for this, e.g. `LoaderTest.testPadding` so I added this logic to pass those tests.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 3 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:89]: Yes, that's a good simplification.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 4 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:104]: Why do you think it's redundant? We could potentially go backwards forever if the UTF8 is corrupt. Assuming valid UTF8, the mismatch sequences will either have the same length (e.g. 4 bytes starting with 0xFX, with a mismatch in any of the next 3 bytes, or mismatching ASCII characters), so the loops could be fused, but this would be less resilient to corrupt data. The maximum length for either of these loops is 4 iterations, and they access different buffers, so the performance impact that can be made here is minimal.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 5 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:125]: Yes, this is a good observation. The `String` class mostly takes care of this for us because it will either throw or replace when it encounters an invalid codepoint, and each entry in the dictionary was a string before being UTF8 encoded, as was the parameter, but I think we need to be resilient here in case things change elsewhere. Making the missing values zero would at least ensure consistent ordering for invalid codepoints.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+ */
+package org.apache.pinot.segment.local.io.util;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+
+public class ValueReaderComparisons {
+
+    private ValueReaderComparisons() {
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:27]: I just don't have my IDE set up like yours, and I tweaked the formatting manually until checkstyle passed. Has the project considered auto formatting with spotless as suggested [here](https://github.com/apache/pinot/pull/10000#issuecomment-1355938476)? I use it on other projects and it takes the pain out of maintaining consistent style without synchronising IDEs (not everyone uses IntelliJ) and IDE settings globally.

In the short term I will find the IDE settings and reformat the files I've changed here.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 4 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java]: I've changed this so there's only one loop, having used the `String` class before storage and in the query layer makes this safe enough.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 5 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java]: I've changed this to the same replacement character used by the `String` class to ensure that comparisons are consistent with it.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java]: Agreed. We have created #6790 to track this, but the work is not travail because we need to switch from intellij to eclipse format which is supported by spotless.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+          // check if the stored string continues beyond the length of the parameter
+          return dataBuffer.getByte(startOffset + bytes.length) == 0 ? 0 : 1;
+        }
+      } else {
+        // then we know the length precisely or know that the parameter is at least as long as we can store
+        return length - bytes.length;
+      }
+    }
+    // we know the position of the mismatch but need to do utf8 decoding before comparison
+    // to respect collation rules
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:81]: Could you help add an example in the comment to help understand the need of compareUtf8()? 

Would compareUtf8() return 0? If so, is there need to compare the remaining bytes after the mismatchPosition?

Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:81]: Hi @klsince it's mostly to do with comparing invalid UTF8 with non ASCII text. For example, comparing the strings 
"𠜎" (UTF8=f0a09c8e) and the garbage string which renders as "�ػ" (UTF8=efbfbdd8bb) then there is a mismatch at zero, but 
```java
Byte.compareUnsigned((byte)0xf0, (byte)0xef);
``` 
gives you -1, but 
```java
new String(BytesUtils.toByteArray("f0a09c8e"), StandardCharsets.UTF_8).compareTo(new String(BytesUtils.toByteArray("efbfbdd8bb"));
``` 
agrees with the implementation on this branch that the comparison is 10172, because f0a09c8e is a valid UTF8 sequence and efbfbdd8bb isn't.

As for whether `compareUtf8` can return zero, that's a good question. It could return zero for two invalid UTF8 sequences, so it will need more testing against `String`. Thanks for raising this.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded, byte[] bytes) {
+    // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+    if (mismatchPosition == -1) {
+      if (padded && bytes.length < length) {
+        // need to figure out whether the unpadded string is longer than the parameter or not
+        if (bytes.length == 0) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:65]: IMO we can simplify this logic and change `LoaderTest` to make it pass. If we want to keep the behavior the same in this PR, we can do a separate PR with just this behavior change.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+          // just need nonzero first byte
+          return dataBuffer.getByte(startOffset) == 0 ? 0 : 1;
+        } else if (bytes[bytes.length - 1] == 0) {
+          // we can't store zero except as padding, but can have it in parameters, so check the last byte
+          return dataBuffer.getByte(startOffset + bytes.length - 1) == 0 ? -1 : 0;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:70]: I think we can directly `return -1` here as the last byte is always 0 under this branch?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Thanks for adding the tests!

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 11 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:65]: I'm wary of removing test cases, it _looks_ like a regression test, it looks like @siddharthteotia added it so he should probably be consulted before removing it. Let's not do it on this branch though. FWIW I feel like the number of special cases is just the right side of manageable though, and if we can make this distinguish between zero padded strings and their valid prefixes with one special case, I'm personally inclined to do that.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 12 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:70]: I recall there was a test case failing without this logic, I'll find out which one and simplify the logic if I can't find it.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 10 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java:80]: I just added more testing and I can't make `compareUtf8` produce zero if either the stored data or the parameter were obtained from `String` instance.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java (hunk 12 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java]: I simplified this.

**Comment by Jackie-Jiang** [comment]
> I think @klsince's question needs addressing - what happens when there are two invalid UTF8 sequences at the mismatch, can we return 0 spuriously, I'll address that before merging.

Could that ever happen? All the UTF-8 are encoded from strings.

**Comment by richardstartin** [comment]
> > I think @klsince's question needs addressing - what happens when there are two invalid UTF8 sequences at the mismatch, can we return 0 spuriously, I'll address that before merging.
> 
> Could that ever happen? All the UTF-8 are encoded from strings.

It can’t happen because at the very least the parameter came from the query layer as a string. I can’t think of a way to break this logic except when the behaviour is undefined for the string class itself. When I first implemented a simpler version of this, a lot of test cases elsewhere in Pinot failed, even after my own tests which generated a comprehensive range of UTF8 sequences was passing, so the test suite is quite sensitive to this change, and is passing now.
